File: bindings/java/src/main/java/dev/tachyon_ipc/RxGuard.java
package dev.tachyon_ipc;

import java.lang.foreign.MemorySegment;

/**
 * A zero-copy read projection of a single message from the consumer's SPSC arena.
 *
 * @apiNote The memory segment is strictly read-only and bounds-checked. It must not be
 * accessed after the guard is closed or committed.
 */
public final class RxGuard implements AutoCloseable {

	/**
	 * The native bus pointer used for FFI operations.
	 */
	private final MemorySegment busHandle;

	/**
	 * The immutable memory segment pointing to the payload.
	 */
	private MemorySegment data;

	/**
	 * The user-defined protocol identifier for this message payload.
	 */
	private final int typeId;

	/**
	 * The exact number of valid payload bytes.
	 */
	private final long actualSize;

	/**
	 * State flag indicating whether the message has been acknowledged.
	 */
	private boolean consumed;

	/**
	 * Internal constructor invoked exclusively by the {@link TachyonBus}.
	 *
	 * @param busHandle  The native bus pointer.
	 * @param safePtr    The bounded, read-only memory segment containing the message.
	 * @param typeId     The extracted protocol identifier.
	 * @param actualSize The exact payload size.
	 */
	RxGuard(MemorySegment busHandle, MemorySegment safePtr, int typeId, long actualSize) {
		this.busHandle = busHandle;
		this.typeId = typeId;
		this.actualSize = actualSize;
		this.data = safePtr.asReadOnly();
		this.consumed = false;
	}

	/**
	 * Retrieves the read-only memory segment for the payload.
	 *
	 * @return The immutable memory segment pointing to the payload in the shared memory file.
	 * @throws IllegalStateException If the guard has already been committed or closed.
	 */
	public MemorySegment getData() {
		checkState();
		return data;
	}

	/**
	 * Retrieves the protocol identifier associated with this message.
	 *
	 * @return The integer type ID.
	 */
	public int getTypeId() {
		return typeId;
	}

	/**
	 * Retrieves the actual size of the payload.
	 *
	 * @return The size in bytes.
	 */
	public long getActualSize() {
		return actualSize;
	}

	/**
	 * Acknowledges the processing of the message and advances the ring buffer's consumer head.
	 */
	public void commit() {
		checkState();
		this.consumed = true;
		TachyonABI.commitRx(busHandle);
	}

	/**
	 * Finalizes the read lifecycle.
	 *
	 * @implSpec Automatically commits the read transaction if not explicitly consumed,
	 * preventing consumer-side stalls and freeing cache lines for the producer.
	 * The internal memory segment reference is nullified to strictly prevent use-after-free.
	 */
	@Override
	public void close() {
		try {
			if (!consumed) {
				commit();
			}
		} finally {
			this.data = null;
		}
	}

	/**
	 * Validates the temporal lifecycle of the guard.
	 *
	 * @throws IllegalStateException If the guard has already been committed or closed.
	 */
	private void checkState() {
		if (consumed || data == null) {
			throw new IllegalStateException("RxGuard has already been committed or closed");
		}
	}
}

File: bindings/csharp/src/TachyonIpc/RxGuard.cs
using TachyonIpc.Native;

namespace TachyonIpc;

/// <summary>
/// Read-side guard for a single message, returned by Bus.TryReceive / Bus.ReceiveSpin / Bus.ReceiveBlocking.
/// Must be used with <c>using</c>.
/// Dispose() commits (advances the read pointer).
/// </summary>
/// <remarks>
/// ref struct, stack-only, zero allocation.
/// The underlying shared memory is valid only until Dispose() is called.
/// Never store a reference to <see cref="Data"/> beyond the guard's lifetime.
/// </remarks>
public unsafe ref struct RxGuard
{
    private readonly nint _bus;
    private readonly byte* _ptr;
    private readonly nuint _actualSize;
    private bool _done;

    /// <summary>bits [15:0] = msg_type, bits [31:16] = route_id.</summary>
    public readonly uint TypeId;

    /// <summary>Decoded msg_type component of <see cref="TypeId"/>.</summary>
    public ushort MsgType => TachyonIpc.TypeId.MsgType(TypeId);

    /// <summary>Decoded route_id component of <see cref="TypeId"/>.</summary>
    public ushort RouteId => TachyonIpc.TypeId.RouteId(TypeId);

    internal RxGuard(nint bus, void* ptr, nuint actualSize, uint typeId)
    {
        _bus = bus;
        _ptr = (byte*)ptr;
        _actualSize = actualSize;
        _done = false;
        TypeId = typeId;
    }

    /// <summary>
    /// Read-only view of the message payload.
    /// Valid only within the guard's scope (do not retain after Dispose()).
    /// </summary>
    public ReadOnlySpan<byte> Data
    {
        get
        {
            if (_done) ThrowInvalidated();
            return new ReadOnlySpan<byte>(_ptr, checked((int)_actualSize));
        }
    }

    /// <summary>Advances the read pointer. Called automatically by Dispose.</summary>
    public void Commit()
    {
        if (_done) return;
        TachyonException.ThrowIfError(
            TachyonNative.tachyon_commit_rx(_bus),
            nameof(TachyonNative.tachyon_commit_rx));
        _done = true;
    }

    /// <summary>Commits and releases the slot.</summary>
    public void Dispose() => Commit();

    private static void ThrowInvalidated() =>
        throw new InvalidOperationException("RxGuard has already been committed.");
}